### PR TITLE
spatialGradient() reworked to use wide universal intrinsics

### PR DESCRIPTION
**Pull request by terfendail**
### This pullrequest changes
spatialGradient() reworked to use wide universal intrinsics

```
buildworker:Win64 OpenCL=windows-2
```

### CODE REVIEW

**Comment by terfendail** [comment]
<details>
<summary>Performance for SSE2 baseline</summary>

| Performance test | Reference time |  PR time | Speedup |
|------------------|----------------|----------|---------|
|spatialGradient::Size_Ksize_BorderType::(640x360, 3, 1)|0.070|0.069|1.01|
|spatialGradient::Size_Ksize_BorderType::(640x360, 3, 4)|0.067|0.067|0.99|
|spatialGradient::Size_Ksize_BorderType::(960x540, 3, 1)|0.146|0.156|0.94|
|spatialGradient::Size_Ksize_BorderType::(960x540, 3, 4)|0.145|0.155|0.94|
|spatialGradient::Size_Ksize_BorderType::(1280x720, 3, 1)|0.263|0.268|0.98|
|spatialGradient::Size_Ksize_BorderType::(1280x720, 3, 4)|0.254|0.271|0.94|
|spatialGradient::Size_Ksize_BorderType::(1920x1080, 3, 1)|0.582|0.599|0.97|
|spatialGradient::Size_Ksize_BorderType::(1920x1080, 3, 4)|0.567|0.591|0.96|
</details>
<details>
<summary>Performance for SSE4_2 baseline</summary>

| Performance test | Reference time |  PR time | Speedup |
|------------------|----------------|----------|---------|
|spatialGradient::Size_Ksize_BorderType::(640x360, 3, 1)|0.068|0.066|1.03|
|spatialGradient::Size_Ksize_BorderType::(640x360, 3, 4)|0.067|0.065|1.03|
|spatialGradient::Size_Ksize_BorderType::(960x540, 3, 1)|0.147|0.146|1.00|
|spatialGradient::Size_Ksize_BorderType::(960x540, 3, 4)|0.146|0.148|0.98|
|spatialGradient::Size_Ksize_BorderType::(1280x720, 3, 1)|0.245|0.254|0.96|
|spatialGradient::Size_Ksize_BorderType::(1280x720, 3, 4)|0.244|0.257|0.95|
|spatialGradient::Size_Ksize_BorderType::(1920x1080, 3, 1)|0.578|0.566|1.02|
|spatialGradient::Size_Ksize_BorderType::(1920x1080, 3, 4)|0.544|0.566|0.96|
</details>
<details>
<summary>Performance for AVX2 baseline</summary>

| Performance test | Reference time |  PR time | Speedup |
|------------------|----------------|----------|---------|
|spatialGradient::Size_Ksize_BorderType::(640x360, 3, 1)|0.066|0.068|0.96|
|spatialGradient::Size_Ksize_BorderType::(640x360, 3, 4)|0.064|0.067|0.96|
|spatialGradient::Size_Ksize_BorderType::(960x540, 3, 1)|0.146|0.143|1.02|
|spatialGradient::Size_Ksize_BorderType::(960x540, 3, 4)|0.149|0.145|1.02|
|spatialGradient::Size_Ksize_BorderType::(1280x720, 3, 1)|0.256|0.239|1.07|
|spatialGradient::Size_Ksize_BorderType::(1280x720, 3, 4)|0.259|0.240|1.08|
|spatialGradient::Size_Ksize_BorderType::(1920x1080, 3, 1)|0.584|0.523|1.12|
|spatialGradient::Size_Ksize_BorderType::(1920x1080, 3, 4)|0.585|0.520|1.12|
</details>

File: modules/imgproc/src/spatialgradient.cpp [hunk 1]
```diff
@@ -129,133 +129,130 @@ void spatialGradient( InputArray _src, OutputArray _dx, OutputArray _dy,
 
     int i_start = 0;
     int j_start = 0;
-#if CV_SIMD128
-    if(hasSIMD128())
+#if CV_SIMD
+    uchar *m_src;
+    short *n_dx, *n_dy;
```
Comment (alalek) [inline modules/imgproc/src/spatialgradient.cpp]: Do we need these variables outside of the for loop below?

Comment (terfendail) [reply modules/imgproc/src/spatialgradient.cpp]: Moved variables inside loops

File: modules/imgproc/src/spatialgradient.cpp [hunk 2]
```diff
@@ -123,139 +123,131 @@ void spatialGradient( InputArray _src, OutputArray _dx, OutputArray _dy,
...
+    // [row][column]
+    // m: offset -1
+    // n: offset  0
+    // p: offset  1
+    // Example: umn is offset -1 in row and offset 0 in column
+    for ( i = 0; i < H - 1; i += 2 )
     {
+        uchar *p_src;//previous row
+        if   ( i == 0 ) p_src = src.ptr<uchar>(i_top);
+        else            p_src = src.ptr<uchar>(i-1);
```
Comment (alalek) [inline modules/imgproc/src/spatialgradient.cpp]:
```
const uchar *p_src = src.ptr<uchar>(( i == 0 ) ? i_top : i - 1);
```

Similar for `m_src`.

Comment (terfendail) [reply modules/imgproc/src/spatialgradient.cpp]: fixed

**Pull request merged**